design/design-doc-external-secret-stores.md

```diff
@@ -0,0 +1,416 @@
+approach in this sense.
+
+In this document we will investigate alternative options to support additional
+secret stores in Crossplane. Throughout the document, we will mostly focus on
+[Vault] as the most popular secret store today.
+
+## Goals
+
+The goal of this document is to investigate alternative approaches to support
+external secret stores in Crossplane to store connection details. The solution
+should apply to both Crossplane itself (for composite connection details) and
+providers (for managed resource connection details).
+
+## Out of Scope
+
+This document focuses on writing/publishing connection details to external
+stores and not about reading/consuming them in the form of provider credentials
+or to consume via the workloads. There are already ways to consume secrets in
+external stores from Kubernetes and is out of the scope for this document.
+
+See [Vault credential injection guide] to see how one can configure Vault and
+Crossplane to consume provider credentials from Vault. To consume secrets living
+inside Vault from workloads running in Kubernetes, one can consider
+[injecting via Vault agent] or use solutions like [Vault secrets operator].
+
+### Design
+
+There are a couple of alternative approaches that we would affect the design.
+We would like to list them all here and finally agree on one of them.
+
+#### Option 1: Supporting Additional Stores in Crossplane-runtime
+
+This is the most straightforward approach. We will implement support for writing
+to external stores in _crossplane-runtime_ via importing and using their 
+clients. With this solution, Crossplane and providers would immediately get
+this feature when configured properly.
+
+![Secret Store Runtime][secret-store-runtime]
+
+##### API
+
+We will deprecate existing `writeConnectionSecretToRef` field in favor of a
+`publishConnectionDetailsTo` field which would support configuring multiple
+secret backends. This will apply to not only managed resources but also to
+_Composite Resources_ and _Composite Resource Claims_.
+
+```
+apiVersion: widgets.crossplane.io/v1beta1
+kind: Example
+metadata:
+  name: my-cool-managed-resource
+spec:
+  forProvider:
+    coolness: 24
+  publishConnectionDetailsTo:
+  - store: KubernetesSecret
+    kubernetesSecret:
+      secretRef:
+        name: my-cool-managed-resource-connection
+        namespace: secrets
+  - store: Vault
+    vault:
+      path: /path/to/secret/my-cool-managed-resource-connection
+      configRef:
+        name: vault-primary
+```
+
+For compositions, we will similarly deprecate existing 
+`writeConnectionSecretsToNamespace` in favor of the same 
+`publishConnectionDetailsTo` field. This would be used for any partial input
+that needs to be enforced/provided by the composition authors.
+
+```
+apiVersion: apiextensions.crossplane.io/v1
+kind: Composition
+metadata:
+  name: compositeclusters.aws.platformref.crossplane.io
+spec:
```

Comment:
Comment from @negz 

The writeConnectionSecretsToNamespace field on a Composition today controls which namespace XRs (that use that Composition) will write secrets to. So it's a partial input to the XR's writeConnectionSecretsToRef. We needed to add this because we wanted platform operators, not claim authors, to choose which namespace the XR secret was written to.

I'm not sure this makes sense in the Vault context. In fact we probably need to think more about the relationship between MR, XR, and XRC connection details when vault is in use. Do we need to write all three like we do when we use Secrets?

Reply:
I think we could still pass some partial input in the Vault context like which config to use or even a parent path to use in Vault which could be enforced by platform operators. This would however require us to adapt our interface in a way that we get `parentPath` as a separate argument.

 - type: Vault 
   vault:
     parentPath: /path/to/parent
     name: my-cool-managed-resource-connection 
     configRef: 
       name: vault-primary


However, I am not sure how this is applicable for other secret stores that we would want to support in the feature.

Reply:
Based on the description in the `Configuring External Secret Stores` section, we might be double-configuring the secret store at this point.

If we assume that SecretStores will always be enabled/disabled by platform operators using a StoreConfig, then we need not identify the store type when we access the secret store.  We could do something like this for vault:

```
apiVersion: secrets.crossplane.io/v1alpha1
kind: StoreConfig
metadata:
  name: vault-primary
Spec:
  type: Vault
  credentials:
    source: Secret
    secretRef:
      namespace: crossplane-system
      name: vault-creds
      key: config
```

Then write to secrets with:
```
publishConnectionDetailsTo:
- storeName: vault-primary
  secretName: /path/to/my/secrets/my-secrets
```

Or, with AWS Secrets Manager:
```
apiVersion: secrets.crossplane.io/v1alpha1
kind: StoreConfig
metadata:
  name: aws-secrets-primary
Spec:
  type: AWSSecret
  credentials:
    source: Secret
    secretRef:
      namespace: crossplane-system
      name: aws-creds
      key: config
```
Leveraged by:
```
publishConnectionDetailsTo:
- storeName: aws-secrets-primary
  secretName: my-secrets
```

Kubernetes could be supported as a `default` storeConfig which is always a kubernetes secret.
```
publishConnectionDetailsTo:
- storeName: default
  secretNamespace: claim-namespaces
  secretName: my-secrets
```